Python/ceval.c

```diff
@@ -863,6 +863,75 @@ static const binaryfunc binary_ops[] = {
+        if (-_PY_NSMALLNEGINTS <= i && i < _PY_NSMALLPOSINTS) {                \
+            Py_DECREF(lhs);                                                    \
+            PyLongObject *res = &_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS + i];    \
+            Py_INCREF(res);                                                    \
+            SET_TOP((PyObject *)res);                                          \
+            DISPATCH();                                                        \
+        }                                                                      \
+        bool inplace = NEXT_OP_STORES(lhs);                                    \
+        if (Py_ABS(i) < PyLong_BASE && Py_REFCNT(lhs) == inplace + 1) {        \
+            assert(lhs < (PyObject *)&_PyLong_SMALL_INTS[0] ||                 \
+                   lhs >= (PyObject *)&_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS +  \
+                                                          _PY_NSMALLPOSINTS]); \
+            ((PyLongObject *)lhs)->ob_digit[0] = Py_ABS(i);                    \
+            Py_SET_SIZE(lhs, i < 0 ? -1 : 1);                                  \
+            DISPATCH();                                                        \
+        }                                                                      \
+        Py_DECREF(lhs);                                                        \
+        PyObject *res = PyLong_FromLong(i);                                    \
+        SET_TOP(res);                                                          \
+        if (res == NULL) {                                                     \
+            goto error;                                                        \
+        }                                                                      \
+        DISPATCH();                                                            \
+    } while (0)
+
+#define BINARY_OP_FAST_FLOAT(OP)                       \
+    do {                                               \
+        PyObject *lhs = SECOND();                      \
+        PyObject *rhs = TOP();                         \
+        DEOPT_IF(!PyFloat_CheckExact(lhs), BINARY_OP); \
+        DEOPT_IF(!PyFloat_CheckExact(rhs), BINARY_OP); \
+        STAT_INC(BINARY_OP, hit);                      \
+        double l = PyFloat_AS_DOUBLE(lhs);             \
+        double r = PyFloat_AS_DOUBLE(rhs);             \
+        double d = l OP r;                             \
+        Py_DECREF(rhs);                                \
+        STACK_SHRINK(1);                               \
+        bool inplace = NEXT_OP_STORES(lhs);            \
+        if (Py_REFCNT(lhs) == inplace + 1) {           \
+            PyFloat_AS_DOUBLE(lhs) = d;                \
+            DISPATCH();                                \
+        }                                              \
+        Py_DECREF(lhs);                                \
```

Comment:
Can this lead to a use-after-free in some strange situation? Would it be safer to move this DECREF a couple of lines down?

Reply:
Hm, I don't see how. We are only `Py_DECREF`'ing `float` objects (so no chance failing/unwinding), and we `SET_TOP(res)` before checking for error.

I can move it if it seems clearer to you, though.

Reply:
Ah, my mistake -- no need.